src/understand-tidb/session.md

```diff
@@ -1 +1,122 @@
+	}
+	return nil
+}},
+```
+
+In the above example, skipping the `SetSession` function is useful because it prevents the `CollationDatabase` from being overwritten when the session is initialized. This is only expected if the user issues a statement to change the `CharsetDatabase` value.
+
+### Differences from MySQL
+
+In TiDB, changes to `GLOBAL` scoped system variables are propagated to other TiDB servers in the cluster and persist across restarts. The notification event to other servers is sent via an etcd channel in the call `domain.GetDomain(s).NotifyUpdateSysVarCache()`:
+
+```go
+// replaceGlobalVariablesTableValue executes restricted sql updates the variable value
+// It will then notify the etcd channel that the value has changed.
+func (s *session) replaceGlobalVariablesTableValue(ctx context.Context, varName, val string) error {
+	stmt, err := s.ParseWithParams(ctx, `REPLACE INTO %n.%n (variable_name, variable_value) VALUES (%?, %?)`, mysql.SystemDB, mysql.GlobalVariablesTable, varName, val)
+	if err != nil {
+		return err
+	}
+	_, _, err = s.ExecRestrictedStmt(ctx, stmt)
+	domain.GetDomain(s).NotifyUpdateSysVarCache() // <-- the notification happens here
+	return err
+}
+```
+
+Because `GLOBAL` scoped variables are propagated to other servers, TiDB also has a special concept of "instance-scoped variables". An instance scoped variable is actually a `SESSION` scoped variable that has a `GetSession` method which returns data that is specific to an instance. For example, [`tidb_general_log`](https://github.com/pingcap/tidb/blob/246c59d5c926780235dd25aef27c469ffe376e21/sessionctx/variable/sysvar.go#L1315-L1321):
+
+```go
+{Scope: ScopeSession, Name: TiDBGeneralLog, Value: BoolToOnOff(DefTiDBGeneralLog), Type: TypeBool, skipInit: true, SetSession: func(s *SessionVars, val string) error {
+	ProcessGeneralLog.Store(TiDBOptOn(val))
+	return nil
+}, GetSession: func(s *SessionVars) (string, error) {
+	return BoolToOnOff(ProcessGeneralLog.Load()), nil
+}},
+```
+
+The decision to make an option such as `tidb_general_log` instance scoped is because it references a file on the local filesystem. This may create issues when global, as the path may not be writable on each tidb-server in the cluster.
+
+As you can see by the `Scope: Session`, instance-scoped variables are not natively handled by the sysvar framework, but are instead denoted by the `GetSession()` function reading from a global location. The documentation for [`tidb_general_log`](https://docs.pingcap.com/tidb/dev/system-variables#tidb_general_log) also notes it as "instance" scoped by convention.
+
+## Transaction state
+
+The `session` is responsible for keeping modified KV-pairs in an in-memory buffer until the transaction commits. A `commit` statement only sets the session variable state that it is [no longer in an active transaction](https://github.com/pingcap/tidb/blob/246c59d5c926780235dd25aef27c469ffe376e21/executor/simple.go#L701-L703):
```

Comment:
```suggestion
The `session` is responsible for keeping modified key-value pairs in an in-memory buffer until the transaction commits. A `commit` statement only sets the session variable state that it is [no longer in an active transaction](https://github.com/pingcap/tidb/blob/246c59d5c926780235dd25aef27c469ffe376e21/executor/simple.go#L701-L703):
```

Reply:
where is the "in-memory buffer"? do you abstract it an interface?

Reply:
It's in the session struct. `s.txn` is a LazyTxn. I've added some clarifications, PTAL.